pkg/config/protocol_config.go

```diff
@@ -55,26 +57,124 @@ type (
 		// Whether to verify transactions in received blocks.
 		VerifyTransactions bool `yaml:"VerifyTransactions"`
 	}
 )
 
+// heightNumber is an auxiliary structure for configuration checks.
+type heightNumber struct {
+	h uint32
+	n int
+}
+
 // Validate checks ProtocolConfiguration for internal consistency and returns
 // error if anything inappropriate found. Other methods can rely on protocol
 // validity after this.
 func (p *ProtocolConfiguration) Validate() error {
-	if len(p.StandbyCommittee) < p.ValidatorsCount {
-		return errors.New("validators count can be less than the size of StandbyCommittee")
-	}
+	var err error
+
 	for name := range p.NativeUpdateHistories {
 		if !nativenames.IsValid(name) {
 			return fmt.Errorf("NativeActivations configuration section contains unexpected native contract name: %s", name)
 		}
 	}
+	if p.ValidatorsCount != 0 && len(p.ValidatorsHistory) != 0 {
+		return errors.New("configuration should either have ValidatorsCount or ValidatorsHistory, not both")
+	}
+	if len(p.StandbyCommittee) < p.ValidatorsCount {
+		return errors.New("validators count can be less than the size of StandbyCommittee")
+	}
+	var arr = make([]heightNumber, 0, len(p.CommitteeHistory))
+	for h, n := range p.CommitteeHistory {
+		if int(n) > len(p.StandbyCommittee) {
+			return errors.New("too small committee for given number of members")
+		}
+		arr = append(arr, heightNumber{h, n})
+	}
+	if len(arr) != 0 {
+		err = sortCheckZero(arr, "CommitteeHistory")
+		if err != nil {
+			return err
+		}
+		for i, hn := range arr[1:] {
+			if int64(hn.h)%int64(hn.n) != 0 || int64(hn.h)%int64(arr[i].n) != 0 {
+				return fmt.Errorf("invalid CommitteeHistory: bad %d height for %d and %d committee", hn.h, hn.n, arr[i].n)
+			}
+		}
+	}
+	arr = arr[:0]
+	for h, n := range p.ValidatorsHistory {
+		if int(n) > len(p.StandbyCommittee) {
+			return errors.New("too small committee for given number of validators")
+		}
+		arr = append(arr, heightNumber{h, n})
+	}
+	if len(arr) != 0 {
+		err = sortCheckZero(arr, "ValidatorsHistory")
+		if err != nil {
+			return err
+		}
+		for _, hn := range arr {
+			if int64(hn.n) > int64(p.GetCommitteeSize(hn.h)) {
+				return fmt.Errorf("requested number of validators is too big: %d at %d", hn.n, hn.h)
+			}
+			if int64(hn.h)%int64(p.GetCommitteeSize(hn.h)) != 0 {
+				return fmt.Errorf("validators number change is not aligned with committee change at %d", hn.h)
+			}
+		}
+	}
 	return nil
 }
+
+// sortCheckZero sorts heightNumber array and checks for zero height presence.
+func sortCheckZero(arr []heightNumber, field string) error {
+	sort.Slice(arr, func(i, j int) bool {
+		return arr[i].h < arr[j].h
+	})
+	if arr[0].h != 0 {
+		return fmt.Errorf("invalid %s: no height 0 specified", field)
+	}
+	return nil
+}
+
+// GetCommitteeSize returns the committee size for the given height. It implies
+// valid configuration file.
+func (p *ProtocolConfiguration) GetCommitteeSize(height uint32) int {
```

Comment:
It's supposed to be small, but I understand your concern. Map is very convenient to use in YAML, I'm not sure sure about an array of structures (which it effectively will be in this case).

Reply:
I thought about creating an auxiliary structure for YAML unmarshalling, so that we can keep it as map in YAML and as array in the resulting config structure. But if it's small, then it's OK.

Reply:
Maybe we can cache them in `Validate()`, but I'm still not sure it's worth that.

Reply:
Let's stay with the map.